### DISCUSSION TITLE
Mono-MultiTasking

### DISCUSSION THREAD

**Discussion by Benis67** [RomanKubiak/ctrlr#360] [General]
I'm making an editor for Roland ZenCore synths (Fantom, Ax Edge, JupiterX / Xm) and I have a problem that I don't know if it can be solved.
For the analysis of the midi messages received by PC from synth I use the function
midiMessageReceived = function (midiMessage) which is called in the main panel in "Called when a a panel receives a MIDI Message.

I'm creating UI buttons that call functions through the classic "Called when the Modulator value Change".
Let's take for example where a UI buttons uses a function that does two operations:

1) asks the synth to send one of the SYSEX midi messages
2) stores data on file

the problem I have is that the order I would like is not respected, that is:
a) I ask the synth to send SYSEX midi messages
b) the synth sends the messages and midiMessageReceived = function (midiMessage) performs its task
c) I store the data

what happens is that after I ask the synth to send one of the SYSEX midi messages, the data is first stored on file and only after function is finished , the midiMessageReceived = function (midiMessage) does its task.

Is there no way to get the requested a-b-c order?
https://www.youtube.com/watch?v=-p2vKKN25GQ

**Comment by tomlem** [comment]
The youtube link is not working.

Can you post the function that is called when the modulator value changes (asking to send the sysex and to store the data)?

**Reply by Benis67** [reply]
You tube link is now corrected :-)  https://www.youtube.com/watch?v=-p2vKKN25GQ
This is the function that is called when the modulator value changes 

SaveToneToFile = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
if panel:getBootstrapState() then
        return
end
if panel:getRestoreState() == true or panel:getProgramState() == true then
        return
end

local mStr=string.format("F0 41 10 00 00 00 52 11 01 01 00 00 00 00 20 7F 00 F7")  -- request of sysex tone data 
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

memB=MemoryBlock()

memB=MemoryBlock(ToneCommonALL)
memB:append(ToneMFXALL)
memB:append(TonePMTALL)
memB:append(TonePartial1ALL)
-- other part--
memB:append(ToneSynthCommonALL)
memB:append(ToneSynthPMTALL)
memB:append(ToneSynthPartia1ALL)
memB:append(ToneSynthPartia2ALL)
memB:append(ToneSynthPartia3ALL)
memB:append(ToneSynthPartia4ALL)

saveToFile(memB)
end 

function saveToFile(dataToSave) -- save memoryBlock to .syx file
    file = utils.saveFileWindow("Save file", File.getSpecialLocation(File.userDesktopDirectory), "*.syx", true)
    if file:hasWriteAccess() then
        file:replaceWithData(dataToSave)
 		panel:getLabel("LSTATUSBAR"):setText("Saved \187 " .. memB:getSize() .. " bytes to sysex file ")
    end
end

**Comment by tomlem** [comment]
It seems to me that the synth doesn't get enough time to send its response (several files!).
What if you insert this line after panel:sendMidiMessageNow(CtrlrMidiMessage(mStr)):
`os.execute(sleep(225))
`
(of course you can adapt the sleeptime if needed)?

Another solution may be to set up a timer.
In Discussion #318 _Ctrlr in 60 lessons_ there is a lesson (06) about setting up timers.

**Comment by Benis67** [comment]
Thanks for your answer. The pause was the first thing I did and it showed me that the main panel function midiMessageReceived = function (midiMessage) starts operating only after the funcion SaveToneToFile () ends.
In practice if after

local mStr = string.format ("F0 41 10 00 00 00 52 11 01 01 00 00 00 00 00 20 7F 00 F7") -- request of sysex tone data
panel: sendMidiMessageNow (CtrlrMidiMessage (mStr))

I insert a 10 second pause, the synth immediately receives the request and sends its sysex message with the patch data, but this data remains in some buffer.
Only then at the end of the SaveToneToFile () function (10 seconds later :-( ) such data are processed by midiMessageReceived = function (midiMessage)
What would be needed is to force the intervention of midiMessageReceived = function (midiMessage) of the main panel before closing the SaveToneToFile () function

**Comment by tomlem** [comment]
What version of Ctrlr are you using? On which operating system?

**Reply by Benis67** [reply]
6.04 with windows 10 64 bit

**Comment by tomlem** [comment]
Dnaldoog created a panel specifically for dumping, writing to disk and restoring to synth sysex data. It's named GDSR and it's in #318 _Ctrlr in 60 lessons_ (AttachedFIles map). Interesting to look at.

The problem with the sleep function may be dat it also puts the MidiMessageReceive function to sleep and thus no time can be gained in receiving the midi message.

Remains: a timer or the GDSR system with a specially created intermediate buffer (memoryBlock).

Maybe someone knows another solution?

**Reply by Benis67** [reply]
I know GSDR from which I have taken a very useful piece of code that processes a multiple sysex, breaking it into n single sysex messages that sends to synth. Unfortunately, there is no such situation in that panel.
I solved it by forcing through the graphical interface the succession of several operations that must then be performed by those who use the software.
http://www.benis.it/varie/Step.png

**Comment by tomlem** [comment]
An acceptable and safe solution that also has the advantage of giving the user insight into the structure of the sysex files :-)

Nice panel!

**Reply by Benis67** [reply]
I hope there is also a "software only" method to force the order. maybe someone knows how :-)

**Reply by dnaldoog** [reply]
@Benis67 

It looks like you've done a huge amount of work in those panels!

Wouldn't it be better to put the following in the _midMessageReceived()_ function?

<pre>

local memB=MemoryBlock(ToneCommonALL)
memB:append(ToneMFXALL)
memB:append(TonePMTALL)
memB:append(TonePartial1ALL)
-- other part--
memB:append(ToneSynthCommonALL)
memB:append(ToneSynthPMTALL)
memB:append(ToneSynthPartia1ALL)
memB:append(ToneSynthPartia2ALL)
memB:append(ToneSynthPartia3ALL)
memB:append(ToneSynthPartia4ALL)

saveToFile(memB)
</pre>

so then you send the midi request to the synth with:

<pre>
SaveToneToFile = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
local mStr="F0 41 10 00 00 00 52 11 01 01 00 00 00 00 20 7F 00 F7" -- request of sysex tone data
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

end
</pre>

That should give you:

a) I ask the synth to send SYSEX midi messages
b) the synth sends the messages and midiMessageReceived = function (midiMessage) performs its task
c) I store the data

:)

**Reply by Benis67** [reply]
Thank you. The method you recommend is the only one possible. The problem arises from the fact that the midiMessageReceived = function (midiMessage) function is very complex (it analyzes 33 types of messages and converts all the parameters of a tone, more than 500, into the corresponding modulators) and the problem in the example I have in more functions.
I should use global variables with status "true" "false" and through these to make midiMessageReceived = function (midiMessage) perform operations only in very specific cases.

**Comment by dobo365** [comment]
Hi, sorry for not answering earlier... I have faced this and similar issues since long and @dnaldoog gave you the correct way of doing things.
As you experienced, Ctrlr is not working like some language where you only go to the next line once the current one, being a function, terminated. 
In other words, when in the current function you branch to another function, the current one continues and you should not expect the result of the branch in it.
In you case, a typical one that I use all the time, you should indeed not expect the result of MidiMessageReceived in the function used to call it.
Therefore the order is indeed:
- the user presses a button to get something from the synth. This function ends
- the panel is constantly watching for Midi reception (MidiMessageReceived). When MidiMessageReceived catches something, you need to verify the message (based on size and expected header) and then act upon it. In your case, if size=xxx and header=yyy then call the method to Save the received data in a memoryblock

It is not good to use os.sleep(). Unfortunately, I don't remember the reason but it can cause Ctrlr to hang or crash.
Good luck with your panel. Keep the good work!

**Reply by dnaldoog** [reply]
I was just going to reply with the exact same answer @dobo365 gave. You can filter incoming MIDI by `if midi:getSize()== 32 then ...`  or if you have the same exact message size coming in for different requests with a size of 32 for example,  you could find a unique byte signature like `if midi:getSize()== 32 and midi:getData():getByte(7)==0x3F then ... elseif midi:getSize()== 32 and midi:getData():getByte(7)==0x3E then ... `  -  so you can filter out your 33 different incoming messages with a series of if/else statements.

**Comment by Godlike-Productions** [comment]
OS Sleep is a pain, because it stops everything. Timers are better.

I use flags a lot in these situations.  Set a flag to true when something is received and use an if statement to decide whether to execute it, often in conjunction with the midi receive checks.

**Reply by Benis67** [reply]
thanks for the answer: I too used the tags method. It is more complicated but the desired results are almost always obtained
